akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala

```diff
@@ -1125,7 +1127,18 @@ private[stream] final class RecoverWith[T, M](pf: PartialFunction[Throwable, Gra
       override def onPull(): Unit = pull(in)
     })
 
-    def onFailure(ex: Throwable) = if (pf.isDefinedAt(ex)) switchTo(pf(ex)) else failStage(ex)
+    def onFailure(ex: Throwable) =
+      if (attempt != r && pf.isDefinedAt(ex)) {
```

Comment:
while it is not very likely that attempt overflows to be `-1` I would still rather prefer to check for `-1` explicitly to disable checks and always retry. Also, you should name the constant to something like `InfiniteRetries`.

i.e.
```
if ((maximumRetries == InfiniteRetries || attempt < maximumRetries) && pf.isDefinedAt(ex))
```

then you don't need the extra check `(r > 0)`, just always increase the `attempt` it does not matter anyway.